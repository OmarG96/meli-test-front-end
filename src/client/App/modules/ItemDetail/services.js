import { http } from "../../../utils/http";

export const getItemDetail = async (itemId) => {
  const response = await http.get(`/api/items/${itemId}`);

  return response.data;
};
