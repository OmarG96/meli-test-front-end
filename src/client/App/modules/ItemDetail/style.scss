@import "../../../utils/scss/variables.scss";

.item-detail-root {
  display: flex;

  .column-1 {
    width: fit-content;
    display: flex;
    flex-direction: column;
  }

  .column-2 {
    display: flex;
    flex-direction: column;
    height: fit-content;
    padding: 32px 32px 0 0;
  }

  .picture {
    margin: 32px;
    position: relative;
    display: block;
    width: 680px;
    height: 468px;
    border-radius: 4px;

    img {
      display: block;
      object-fit: contain;
      width: 100%;
      height: 100%;
    }
  }

  .detail {
    display: flex;
    flex-direction: column;
    height: fit-content;
  }

  .condition {
    font-size: 14px;
    color: $gray-6;
    margin-bottom: 16px;
  }

  .title {
    font-size: 24px;
    color: $gray-3;
    font-weight: bold;
  }

  .price {
    font-size: 46px;
    color: $gray-3;
    margin: 32px 0;
    display: flex;
  }

  .buy-button {
    background-color: $blue;
    color: $white;
    border: none;
    cursor: pointer;
    padding: 16px;
    border-radius: 4px;
  }

  .description-container {
    display: flex;
    flex-direction: column;
    margin: 32px;

    .description-title {
      font-size: 28px;
      color: $gray-3;
      margin-bottom: 32px;
    }

    .description {
      color: $gray-6;
      font-size: 16px;
    }
  }
}
