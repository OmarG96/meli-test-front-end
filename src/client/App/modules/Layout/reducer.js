import {
  SEARCH_FETCH,
  SEARCH_FETCH_SUCCESS,
  SEARCH_FETCH_ERROR,
} from "./actionTypes";

const initialState = {
  search: {
    loading: false,
    data: null,
    error: null,
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case SEARCH_FETCH:
      return {
        ...state,
        search: {
          data: null,
          loading: true,
          error: null,
        },
      };

    case SEARCH_FETCH_SUCCESS:
      const { data } = action.payload;
      return {
        ...state,
        search: {
          data,
          loading: false,
          error: null,
        },
      };

    case SEARCH_FETCH_ERROR:
      const { error } = action.payload;
      return {
        ...state,
        search: {
          ...state.search,
          loading: false,
          error: error,
        },
      };

    default: {
      return state;
    }
  }
};

export default reducer;
export { initialState };
