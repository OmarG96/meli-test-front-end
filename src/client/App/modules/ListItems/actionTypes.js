export const SEARCH_FETCH = "SEARCH_FETCH";
export const SEARCH_FETCH_SUCCESS = "SEARCH_FETCH_SUCCESS";
export const SEARCH_FETCH_ERROR = "SEARCH_FETCH_ERROR";
