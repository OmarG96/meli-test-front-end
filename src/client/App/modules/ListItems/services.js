import { http } from "../../../utils/http";

export const searchItems = async (item) => {
  const response = await http.get("/api/items", {
    params: {
      q: item,
    },
  });

  return response.data;
};
