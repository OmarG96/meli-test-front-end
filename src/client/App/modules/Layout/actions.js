import {
  SEARCH_FETCH,
  SEARCH_FETCH_SUCCESS,
  SEARCH_FETCH_ERROR,
} from "./actionTypes";
import { searchItems } from "./services";

export const onFetchStart = () => ({
  type: SEARCH_FETCH,
});

export const onFetchSuccess = (data) => ({
  type: SEARCH_FETCH_SUCCESS,
  payload: { data },
});

export const onFetchError = (error) => ({
  type: SEARCH_FETCH_ERROR,
  payload: { error },
});

export const onFetch = (item) => async (dispatch) => {
  dispatch(onFetchStart());

  try {
    const data = await searchItems(item);
    dispatch(onFetchSuccess(data));
  } catch (error) {
    dispatch(onFetchError(error));
  }
};
