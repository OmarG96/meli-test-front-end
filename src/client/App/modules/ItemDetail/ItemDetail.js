import React, { useCallback, useEffect, useState } from "react";
import { useSelector } from "react-redux";

import { formatMoney } from "../../../utils/methods/formatMoney";
import { useQueryURL } from "../../../utils/hooks";
import parseCondition from "./utils/parseCondition";
import { getItemDetail } from "./services";

import { Breadcrumb, CardContainer, GridContainer } from "../../../common";

import "./style.scss";

const ItemDetail = () => {
  const [isLoading, setIsLoading] = useState(false);
  const [itemData, setItemData] = useState(null);

  const query = useQueryURL();
  const idItem = query.get("id");

  const { data } = useSelector(({ layout: { search } }) => search);
  const { categories } = data || { categories: [] };

  const fetchItemDetail = useCallback(async () => {
    setIsLoading(true);
    try {
      const response = await getItemDetail(idItem);
      setItemData(response.item);
    } catch (error) {
      console.log(error);
    }
    setIsLoading(false);
  }, [setIsLoading, idItem]);

  useEffect(() => {
    fetchItemDetail();
  }, [fetchItemDetail]);

  return (
    <GridContainer>
      <Breadcrumb path={categories} />
      {!isLoading && itemData ? (
        <CardContainer>
          <div className="item-detail-root">
            <div className="column-1">
              <div className="picture">
                <img src={itemData.picture} />
              </div>
              <div className="description-container">
                <span className="description-title">
                  Descripci√≥n del producto
                </span>
                <span className="description">{itemData.description}</span>
              </div>
            </div>
            <div className="column-2">
              <span className="condition">{`${parseCondition(
                itemData.condition
              )} - ${itemData.sold_quantity} vendidos`}</span>
              <span className="title">{itemData.title}</span>
              <span className="price">
                {formatMoney(itemData.price.amount)}
              </span>
              <button className="buy-button">Comprar</button>
            </div>
          </div>
        </CardContainer>
      ) : null}
    </GridContainer>
  );
};

ItemDetail.propTypes = {};

export default ItemDetail;
