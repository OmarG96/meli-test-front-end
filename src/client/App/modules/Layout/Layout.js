import React, { useCallback } from "react";
import { func, node } from "prop-types";
import { useDispatch } from "react-redux";
import { useHistory } from "react-router-dom";

import { SearchBar, Header } from "../../../common";

import logo from "../../../assets/images/Logo_ML.png";
import "./style.scss";
import { onFetch } from "./actions";

const Layout = ({ children }) => {
  const dispatch = useDispatch();
  const history = useHistory();

  const handleSearch = useCallback(
    (item) => {
      dispatch(onFetch(item));
      history.push("/");
    },
    [dispatch, history]
  );

  const handleClickLogo = useCallback(() => {
    history.push("/");
  }, [history]);

  return (
    <div className="layout">
      <Header>
        <img
          role="button-logo"
          className="header-logo"
          src={logo}
          onClick={handleClickLogo}
        />
        <SearchBar onSearch={handleSearch} />
      </Header>
      {children}
    </div>
  );
};

Layout.propTypes = {
  children: node,
  onSearch: func,
};

export default Layout;
