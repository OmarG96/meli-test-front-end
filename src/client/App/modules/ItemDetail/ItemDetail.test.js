import React from "react";
import renderWithProvider from "../../../utils/tests/renderWithProvider";
import ItemDetail from ".";
import { getItemDetail } from "./services";
import { screen, waitFor } from "@testing-library/react";
import { SEARCH_FETCH_SUCCESS } from "../Layout/actionTypes";

jest.mock("./services", () => ({
  getItemDetail: jest.fn(),
}));

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useLocation: () => ({
    pathname: "http://localhost:3000/detalle?id=EXAMPLE123",
  }),
}));

describe("GIVEN ItemDetail", () => {
  describe("WHEN is rendered", () => {
    it("THEN match to snapshot", async () => {
      getItemDetail.mockResolvedValue({
        item: {
          id: "EXAMPLE123",
          title: "Title title title",
          price: {
            currency: "ARS",
            amount: "123456",
          },
          picture: "url.example.com",
          condition: "new",
          free_shipping: true,
          sold_quantity: 123,
          description: "description description description. Description.",
        },
      });
      const { asFragment } = renderWithProvider(<ItemDetail />, {
        onPrepareStore: (store) =>
          store.dispatch({
            type: SEARCH_FETCH_SUCCESS,
            payload: {
              data: { categories: ["example1", "example2", "example3"] },
            },
          }),
      });

      await screen.findByText("Title title title");
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
