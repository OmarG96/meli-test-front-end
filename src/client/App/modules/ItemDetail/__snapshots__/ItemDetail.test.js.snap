// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GIVEN ItemDetail WHEN is rendered THEN match to snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid-container-root"
  >
    <div
      class="breadcrumb-root"
    >
      <span
        class="breadcrumb-route"
      >
        example1
      </span>
      <span
        class="breadcrumb-route"
      >
        example2
      </span>
      <span
        class="breadcrumb-route"
      >
        example3
      </span>
    </div>
    <div
      class="card-container-root"
    >
      <div
        class="item-detail-root"
      >
        <div
          class="column-1"
        >
          <div
            class="picture"
          >
            <img
              src="url.example.com"
            />
          </div>
          <div
            class="description-container"
          >
            <span
              class="description-title"
            >
              Descripci√≥n del producto
            </span>
            <span
              class="description"
            >
              description description description. Description.
            </span>
          </div>
        </div>
        <div
          class="column-2"
        >
          <span
            class="condition"
          >
            Nuevo - 123 vendidos
          </span>
          <span
            class="title"
          >
            Title title title
          </span>
          <span
            class="price"
          >
            $ 123,456
          </span>
          <button
            class="buy-button"
          >
            Comprar
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
